### ISSUE TITLE
Is there a way to use keras flow_from_directory with distributed training (mirrored strategy)?

### ISSUE THREAD

**Issue by kristofgiber** [tensorflow/tensorflow#32032]
I cannot find any documentation or example code on how to do dynamic file loading with distribute strategies in tensorflow 2.0 when doing distributed training. 

I see many examples using keras flow_from_directory which seems to be very nice because it relies on generator and does scaling etc on the fly so doesn't require loading hundreds of thousands of images at once. But the distribute strategy examples (mirrored strategy etc) all show pre-loading the entire dataset and then using dataset.from_tensor_slices() and loading that to strategy.experimental_distribute_dataset(). This is not feasable with large data that exceeds the memory. I've tried to combine the above dynamic loading method from keras with distributed batching but it looks like there is no trivial way to convert some flow_from_directory output to be compatible with what strategy.experimental_distribute_dataset() expects so these two features are probably not compatible with one another.

Is there some other way to do this with distributed training?

I could probably hand-code dynamic file loading during training but it would be pretty slow and basic compared to what generators and flow offer. I would be very surprised if TF2 didn't include this functionality for distributed training yet.

However, in lack of examples or documentation, assuming this is not yet implemented, I am posting it as feature request. Hopefully I'm wrong and there is a way to do it. Then I would recommend to add that to the distribute strategies documentation as example code.

**Comment by kristofgiber** [comment]
While I'd still be very glad if the very convenient and concise flow_from_directory() with generators could work with TF2 distribute strategies (because in the rest of my project I use TF2 tf.compat.v2 for distributed training), I've been looking for alternatives:

An alternative way of dynamically load and preprocess large datasets is described here:
In lack of full examples, I tried to follow the tutorial here:
https://www.tensorflow.org/tutorials/load_data/images
https://www.tensorflow.org/guide/performance/datasets

The output being the same Dataset format what strategy.experimental_distribute_dataset() method takes as input, it should work. However, the tutorial and documentations are all written for tf.compat.v1 (which is deprecated I think) not tf.compat.v2.

There are commands that don't work any more in tf.compat.v2 (eg. there is no output_shape and output_type attribute for Dataset v2). 

While I can make it run with a few small changes, is it safe to run that code in tf.compat.v2 in TF2?
Or is there some updated examples somewhere?

**Comment by yuefengz** [comment]
I think tf.data should serve your needs. I am not sure about the version problem. @jsimsa probably could comment.

**Comment by jsimsa** [comment]
To load data dynamically using tf.data, you can use [tf.data.Dataset.from_generator](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/data/Dataset#from_generator). This should work seamlessly with single worker distribution strategies. For multi-worker distribution strategies, you will need to use [experimental_dataset_from_function](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/distribute/Strategy#experimental_distribute_datasets_from_function) as automatically sharding the generator dataset across workers is not supported.

**Comment by kristofgiber** [comment]
Thank you for the answers! I needed solution for multi-GPU distribution strategy but I'm not sure if the multi-GPUs offered by Gcloud or Azure, that I've been using, are multi-worker systems or not, I think they are just multiple GPUs and multiple CPUs on a single VM.
Since then I've rewritten my code in tf.keras functional API and then got this working very nicely with the alternative method of 
train_dataset = tf.data.Dataset.from_tensor_slices(((train_images, dummydata),train_images)).shuffle(TRAIN_BUF).repeat().batch(BATCH_SIZE)
and then this directly goes to fit(). This works nicely with distribute strategy.

**Comment by jsimsa** [comment]
Note that "worker" means machine. Multiple GPUs on a single machine are a single worker distribution strategy (e.g. MirroredStrategy). In any case, great to hear that you were able to use tf.data (and in an idiomatic way too) to solve your problem!

**Comment by jvishnuvardhan** [comment]
@kristofgiber Can you please close the issue if this issue was resolved already? Thanks!

**Comment by patricksansom** [comment]
Any chance you can provide a little more detail around your solution

train_dataset = tf.data.Dataset.from_tensor_slices(((train_images, dummydata),train_images)).shuffle(TRAIN_BUF).repeat().batch(BATCH_SIZE)

e.g. How have you set up/loaded train_images?

Thanks

**Comment by Kishanrajput** [comment]
@kristofgiber, could you please provide details of your solution. How you set up train_images? I'm using
tf.keras.preprocessing.image.ImageDataGenerator(rescale=1./255).flow_from_dataframe() 
for CPU and single GPU training but when transferring to multi-GPU training with tf.distribute.MirroredStrategy(), I could not find any documentation on how to distribute dataset among GPUs with flow_from_dataframe(). Any help would be appreciated, Thanks!
